### PR TITLE
[JIT] Autodiff - use more accurate requires_grad info

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #78875
* __->__ #78392

When autodiff is constructing the Gradient object, it looks at the
forward graph and records all the outputs that requires_grad into
df_input_vjps. Then at runtime, graph_executor.cpp will detach the
tensors before running the autodiff forward graph, and then add
requires_grad back onto the outputs if they need requires_grad.

Before, the require_grad check was done by just checking
`output->requires_grad()`. But at the point when autodiff is called by
profiling executor, the profiled information is still in the profile
nodes, not on values. So requires_grad would not be set on the output
values, and requires_grad() would default to True on all tensors. As a
result more output tensors than expected would require_grad.

Differential Revision: [D36721504](https://our.internmc.facebook.com/intern/diff/D36721504)

### CODE REVIEW

File: test/jit/test_autodiff.py [hunk 1]
```diff
@@ -49,3 +49,19 @@ def fn(tensor_list: List[torch.Tensor], add_tensor):
         for x in grad_inputs:
             if x is not None:
                 self.assertEqual(0, torch.max(torch.abs(x)).item())
+
+    def test_requires_grad_outputs(self):
```
Comment (davidberard98) [inline test/jit/test_autodiff.py:53]: (this test was failing before the change to autodiff.cpp)

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36721504).

**Comment by davidberard98** [comment]
RUN_TORCHBENCH: ALL

File: torch/csrc/jit/runtime/autodiff.cpp [hunk 2]
```diff
@@ -356,6 +356,28 @@ static Value* createAutogradAdd(Value* a, Value* b) {
   return graph->insertNode(graph->create(prim::AutogradAdd, {a, b}))->output();
 }
 
+namespace {
+bool outputRequiresGrad(Value* output) {
+  if (output->type()->castRaw<TensorType>() == nullptr) {
+    return output->requires_grad();
```
Comment (eellison) [inline torch/csrc/jit/runtime/autodiff.cpp:362]: dont know when we would be putting lists and such to have grad here but is consistent with existing behavior so looks good

File: torch/csrc/jit/runtime/autodiff.cpp [hunk 3]
```diff
@@ -356,6 +356,28 @@ static Value* createAutogradAdd(Value* a, Value* b) {
...
+  }
+  c10::optional<bool> requiresGrad =
+      output->type()->expectRef<TensorType>().requiresGrad();
+  if (requiresGrad.has_value()) {
+    return *requiresGrad;
+  }
+
+  Node* n = output->node();
+  if (n->kind() != prim::profile) {
+    return true;
```
Comment (eellison) [inline torch/csrc/jit/runtime/autodiff.cpp:372]: so this is conservatively saying `true` I guess ?

File: torch/csrc/jit/runtime/autodiff.cpp [hunk 4]
```diff
@@ -356,6 +356,28 @@ static Value* createAutogradAdd(Value* a, Value* b) {
...
+  }
+  if (!n->hasAttribute(attr::profiled_type)) {
```
Comment (eellison) [inline torch/csrc/jit/runtime/autodiff.cpp:374]: We profile on use not on def right, so I guess the output is always `prim::profile`

**Review by eellison — APPROVED** [review]
LGTM - also cc @Krovatkin

File: torch/csrc/jit/runtime/autodiff.cpp (hunk 3 above)
Comment (davidberard98) [reply torch/csrc/jit/runtime/autodiff.cpp:372]: yeah, that's also the default for when the `optional<bool> requiresGrad` is unset in a `TensorType`

File: torch/csrc/jit/runtime/autodiff.cpp (hunk 2 above)
Comment (davidberard98) [reply torch/csrc/jit/runtime/autodiff.cpp:362]: for lists we don't profile so I think we have to assume that they require_grad

File: torch/csrc/jit/runtime/autodiff.cpp (hunk 4 above)
Comment (eellison) [reply torch/csrc/jit/runtime/autodiff.cpp:374]: Could you try something with multiple uses that can't be merged in what group in a test ?

Comment (davidberard98) [reply torch/csrc/jit/runtime/autodiff.cpp:374]: Added fixes in https://github.com/pytorch/pytorch/pull/78875
